packages/validation/src/validators/generic/SyntaxValidator.ts

```diff
@@ -0,0 +1,166 @@
+import {
+    Syntax_ArrayContainsNull_Issue,
+    Syntax_PropertyMissingIssue,
+    Syntax_PropertyNullIssue,
+    Syntax_PropertyTypeIssue,
+    Syntax_PropertyUnknownIssue
+} from "../../issues/index.js"
+import { JsonContext } from "../../json/JsonContext.js"
+import { ValidationResult } from "./ValidationResult.js"
+import { PropertyDefinition, UnknownObjectType } from "./ValidationTypes.js"
+
+/**
+ * Syntax Validator can check whether objects are structurally conforming to the
+ * definitions given in `expectedTypes`.
+ */
+export class SyntaxValidator {
+    validationResult: ValidationResult
+    expectedTypes: Map<string, PropertyDefinition[]>
+
+    constructor(validationResult: ValidationResult, expectedTypes: Map<string, PropertyDefinition[]>) {
+        this.validationResult = validationResult
+        this.expectedTypes = expectedTypes
+    }
+
+    /**
+     * Check whether `obj` is a JSON object that conforms to the definition of `expectedType`.
+     * All errors found will be pushed into the `errors` array, if its length is not 0, the check has failed.
+     * @param obj           The object to validate.
+     * @param expectedType  The expected type of the object.
+     */
+    validate(obj: unknown, expectedType: string) {
+        if (typeof obj !== "object") {
+            throw new Error(`SyntaxValidator.validate: 'obj' is not an object, expected a '${expectedType}'`)
+        }
+        const object = obj as UnknownObjectType
+        const defs = this.expectedTypes.get(expectedType)
+        if (defs === undefined) {
+            throw new Error(`SyntaxValidator.validate: cannot find definition for ${expectedType}`)
+        } else {
+            this.validateObjectProperties(expectedType, defs, object, new JsonContext(null, ["$"]))
+        }
+    }
+
+    /**
+     * Validate whether `object` is structured conform the properties in `propertyDef`
+     * @param originalProperty  The property of which `object` it the value
+     * @param propertyDef       The property definitions that are being validated
+     * @param object            The object being validated
+     * @param jsonContext       The location in the JSON
+     * @private
+     */
+    private validateObjectProperties(originalProperty: string, propertyDef: PropertyDefinition[], object: UnknownObjectType, jsonContext: JsonContext) {
+        if (propertyDef.length === 0) {
+            return
+        }
+        if ((typeof object) !== "object") {
+            this.validationResult.issue(new Syntax_PropertyTypeIssue(jsonContext, originalProperty, "object", typeof object))
+            return
+        }
+        for (const pdef of propertyDef) {
+            const expectedPropertyDefs = this.expectedTypes.get(pdef.expectedType)
+            const validator = pdef.validate!
+            const propertyValue = object[pdef.property]
+            if (propertyValue === undefined) {
+              //  console.log("  missing")
```

Comment:
Please remove commented-out (and uncommented) debugging-style `console.log`s.

Reply:
Right! done.